examples/glium_gl_area/main.rs

```diff
@@ -0,0 +1,280 @@
+    }
+
+    fn get_framebuffer_dimensions(&self) -> (u32, u32) {
+        let area = self.area.upgrade().unwrap();
+        let scale = area.scale_factor();
+        let width = area.width();
+        let height = area.height();
+        ((width * scale) as u32, (height * scale) as u32)
+    }
+
+    fn is_current(&self) -> bool {
+        match self.area.upgrade().unwrap().context() {
+            Some(context) => gdk::GLContext::current() == Some(context),
+            None => false,
+        }
+    }
+
+    unsafe fn make_current(&self) {
+        self.area.upgrade().unwrap().make_current();
+    }
+}
+
+struct GLAreaContext {
+    context: Rc<glium::backend::Context>,
+}
+
+impl glium::backend::Facade for GLAreaContext {
+    fn get_context(&self) -> &Rc<glium::backend::Context> {
+        &self.context
+    }
+}
+
+impl GLAreaContext {
+    /// # Safety
+    ///
+    /// Only call this from GLArea's realize after ensuring there was no error, and don't make any
+    /// modifications to the GL context.
+    unsafe fn new(area: WeakRef<gtk::GLArea>) -> Self {
+        let backend = GLAreaBackend { area };
+        let context = glium::backend::Context::new(backend, true, Default::default()).unwrap();
+        Self { context }
+    }
+
+    fn draw(&self) -> Frame {
+        Frame::new(
+            Rc::clone(&self.context),
+            self.context.get_framebuffer_dimensions(),
+        )
+    }
+}
+
+#[derive(Copy, Clone)]
+struct Vertex {
+    position: [f32; 2],
+    color: [f32; 3],
+}
+
+implement_vertex!(Vertex, position, color);
+
+struct Renderer {
+    context: GLAreaContext,
+    vertex_buffer: VertexBuffer<Vertex>,
+    index_buffer: IndexBuffer<u16>,
+    program: glium::Program,
+}
+
+impl Renderer {
+    /// # Safety
+    ///
+    /// Only call this from GLArea's realize and don't make any modifications to the GL context.
+    unsafe fn new(area: &gtk::GLArea) -> Option<Self> {
+        if area.error().is_some() {
+            return None;
+        }
+
+        let context = GLAreaContext::new(area.downgrade());
+
+        // The following code is from glium's triangle example:
+        // https://github.com/glium/glium/blob/2ff5a35f6b097889c154b42ad0233c6cdc6942f4/examples/triangle.rs
+        let vertex_buffer = VertexBuffer::new(
+            &context,
+            &[
+                Vertex {
+                    position: [-0.5, -0.5],
+                    color: [0., 1., 0.],
+                },
+                Vertex {
+                    position: [0., 0.5],
+                    color: [0., 0., 1.],
+                },
+                Vertex {
+                    position: [0.5, -0.5],
+                    color: [1., 0., 0.],
+                },
+            ],
+        )
+        .unwrap();
+        let index_buffer =
+            IndexBuffer::new(&context, PrimitiveType::TrianglesList, &[0u16, 1, 2]).unwrap();
+        let program = program!(&context,
+            140 => {
+                vertex: "
```

Comment:
Maybe you want to put the shaders in separate files an `include!` them here? That would potentially keep the code easier to read